docs: add docs for check options (#2406)
dequelabs/axe-core: doc/check-options.md

## How Checks Work

[Rules in axe-core](../lib/rules) are made up of individual checks that dictate how the rule works. Each check is typically designed to look for a specific requirement and report back its findings to the rule.

For example, the rule [image-alt](../lib/rules/image-alt.json) uses the checks `has-alt`, `aria-label`, `aria-labelledby`, and `non-empty-title` to determine if the image has an accessible name from an `alt`, `aria-label`, `aria-labelledby`, or `title` attribute (respectively).

Many checks allow you to change how they work through `options` properties. These options can be found in the [checks metadata file](../lib/checks).

For example, the check [has-lang](../lib/checks/language/has-lang.json) takes an `attributes` option which dictates which attributes to check for a lang value.

To customize a checks options, you can use [`axe.configure`](./API.md#api-name-axeconfigure) to configure the check and modify the options as desired.

```js
// configure has-lang check to look at the `hreflang` attribute as well
axe.configure({
	checks: [
		{
			id: 'has-lang',
			options: {
				attributes: ['lang', 'xml:lang', 'hreflang']
			}
		}
	]
});

> ```

Reviewer: something's wonky with the rendered codelisting formatting here:

![image](https://user-images.githubusercontent.com/4044039/88216012-8200ba80-cc2a-11ea-8a41-abaa5fe04caa.png)

Reviewer 2: Aside from it being hard to read (due to the tabs), I don't see anything "wonky" here.

Reviewer: It looks like a linter gone wild thing. The code itself looks fine. Github rendering, not so much.

Reviewer 2: I'm not sure what you're suggesting here. Is there anything that should be "fixed"?

Reviewer: @stephenmathieson: not if the 2 spaces = indentation/prettier stuff is working, no. It's up to Github to render it properly.

Reviewer 2: I do not understand.

### duplicate-img-label

| Option           | Default                                         | Description                                                                           |
| ---------------- | :---------------------------------------------- | :------------------------------------------------------------------------------------ |
| `parentSelector` | `button, [role=button], a[href], p, li, td, th` | Selector used to look for the parent of an image that could duplicate the image label |

### label-content-name-mismatch

| Option               | Default | Description                                                                                                                                                               |
| -------------------- | :------ | :------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| `pixelThreshold`     | `0.1`   | Percent of differences in pixel data or pixel width needed to determine if a font is a ligature font (ligature fonts are ignored when comparing the label to the content) |

> | `occuranceThreshold` | `3`     | Number of times the font is encountered before auto-assigning the font as a ligature or not                                                                               |

Author: Unfortunately that would require a breaking change, sorry this typo wasn't caught before.

### has-lang

| Option       | Default                | Description                         |
| ------------ | :--------------------- | :---------------------------------- |
| `attributes` | `['lang', 'xml:lang']` | Attributes to check for lang values |

### valid-lang

| Option       | Default                | Description                               |
| ------------ | :--------------------- | :---------------------------------------- |
| `attributes` | `['lang', 'xml:lang']` | Attributes to check for valid lang values |

### frame-tested

| Option        | Default | Description                                                |
| ------------- | :------ | :--------------------------------------------------------- |

> | `isViolation` | `false` | If an iframe without axe should be reported as a violation |

Reviewer: ```suggestion
| `isViolation` | `false` | If an `iframe` without axe should be reported as a violation |
```
What does "without axe" mean? an `iframe` that has not been tested by axe? An `iframe` that has not been tested by axe yet, but will be somewhere else?

### no-autoplay-audio

| Option            | Default | Description                                                                   |
| ----------------- | :------ | :---------------------------------------------------------------------------- |

> | `allowedDuration` | `3`     | Maximum time an audio is allowed to autoplay before being marked as violation |

Reviewer: ```suggestion
| `allowedDuration` | `3`     | Maximum time in seconds an audio clip may autoplay before being marked as violation |
```

### css-orientation-lock

| Option            | Default | Description                                                                                                                            |
| ----------------- | :------ | :------------------------------------------------------------------------------------------------------------------------------------- |

> | `degreeThreshold` | `3`     | The difference of degrees from 180 and 90 that are considered orientation lock (for example, 93° is orientation locked but 94° is not) |

Reviewer: Are we talking about a device orientation lock? The web API `ScreenOrientation.lock()`? The W3 Screen Orientation API? Please be specific.

### meta-viewport-large

| Option         | Default | Description                                                                                                |
| -------------- | :------ | :--------------------------------------------------------------------------------------------------------- |

> | `scaleMinimum` | `5`     | The maximum CSS value of `scale-maximum` the check will apply to. Values above this number will be ignored |

Reviewer: ```suggestion
| `scaleMinimum` | `5`     | The `scale-maximum` CSS value of the check applies to. Values above this number will be ignored |
```

> | `scaleMinimum` | `5`     | The maximum CSS value of `scale-maximum` the check will apply to. Values above this number will be ignored |

Reviewer: ```suggestion
| `scaleMinimum` | `5`     | The `scale-minimum` CSS value the check applies to. Values above this number will be ignored |
```

> | `lowerBound`   | `2`     | The minimum CSS value of `scale-minimum` the check will apply to. Values below this number will be ignored |

Reviewer: ```suggestion
| `lowerBound`   | `2`     | The `scale-minimum` CSS value the check applies to. Values below this number will be ignored |
```

### meta-viewport

| Option         | Default | Description                                                                                                |
| -------------- | :------ | :--------------------------------------------------------------------------------------------------------- |

> | `scaleMinimum` | `2`     | The maximum CSS value of `scale-maximum` the check will apply to. Values above this number will be ignored |

Reviewer: ```suggestion
| `scaleMinimum` | `2`     | The `scale-maximum` CSS value the check applies to. Values above this number will be ignored |
```

### header-present

| Option     | Default                                                                                                          | Description                                        |
| ---------- | :--------------------------------------------------------------------------------------------------------------- | :------------------------------------------------- |

> | `selector` | `h1:not([role]), h2:not([role]), h3:not([role]), h4:not([role]), h5:not([role]), h6:not([role]), [role=heading]` | Selector used to determine if a page has a heading |

Reviewer: Hard to read. Same problem as big line of inline CSS above; it's hard to read. Anything we can do to make it clear that this is CSS to make it easier to read is appreciated. (I know this is a CSS selector in theory, but the inline presentation munges stuff together and makes it less obvious that we're talking about a series of very specific selectors here.

### landmark

| Option     | Default             | Description                                                |
| ---------- | :------------------ | :--------------------------------------------------------- |
| `selector` | `main, [role=main]` | Selector used to determine if a page has a landmark region |

### p-as-heading

| Option | Default | Description |
| ------ | :------ | :---------- |

> 

Reviewer: | `margins` | CSS properties {	"weight": 150, "italic": true },	{ "weight": 150,	"size": 1.15	}, { "italic": true, "size": 1.15	}, { "size": 1.4 } | Common CSS values used to display `p` elements as `h1`-`h6` elements determining if a `p` element is being improperly repurposed |

???

### avoid-inline-spacing

| Option          | Default                                             | Description                                      |
| --------------- | :-------------------------------------------------- | :----------------------------------------------- |

> | `cssProperties` | `['line-height', 'letter-spacing', 'word-spacing']` | List of CSS properties to flag as inline spacing |

Reviewer: ```suggestion
| `cssProperties` | `['line-height', 'letter-spacing', 'word-spacing']` | List of inline spacing CSS properties to flag |
```
